assets/contracts/EmotableRepository.sol

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: CC0-1.0
+
+pragma solidity ^0.8.16;
+
+import "./IEmotableRepository.sol";
+import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+
+contract EmotableRepository is IEmotableRepository {
+    // Used to avoid double emoting and control undoing
+    mapping(address => mapping(address => mapping(uint256 => mapping(bytes4 => uint256))))
+        private _emotesUsedByEmoter; // Cheaper than using a bool
+    mapping(address => mapping(uint256 => mapping(bytes4 => uint256)))
+        private _emotesPerToken;
+
+    function emoteCountOf(
+        address collection,
+        uint256 tokenId,
+        bytes4 emoji
+    ) public view returns (uint256) {
+        return _emotesPerToken[collection][tokenId][emoji];
+    }
+
+    function hasEmoterUsedEmote(
+        address emoter,
+        address collection,
+        uint256 tokenId,
+        bytes4 emoji
+    ) public view returns (bool) {
+        return _emotesUsedByEmoter[emoter][collection][tokenId][emoji] == 1;
+    }
+
+     function emote(
+        address collection,
+        uint256 tokenId,
+        bytes4 emoji,
+        bool state
+    ) external override {
+        _emote(collection, tokenId, emoji, state);
+    }
```

Comment:
I guess it's too late now since bytes would change. But since this is a standalone it makes no sense to have an external and an internal function. This will make every operation a bit more expensive I think.

Reply:
It's not really an issue.. I removed the internal function and recalculated the address 👌